[Dynamo] Install module globals per output_graph

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118030
* __->__ #117998

Fixes https://github.com/pytorch/pytorch/issues/117851

In tests, we ran into an issue where:
- In frame A, Dynamo would install a global
- We call reset()
- reset() did not delete the installed global due to a refcycle
- In frame B, Dynamo would re-use the same global
- Python gc ran, deleting the installed global, leading to the compiled
  version of frame B raising NameNotFound

This PR changes the following:
- module globals are now installed at a per-frame basis.
- renames install_global to install_global_unsafe: if the names are not
  unique and end up being re-used across frames, then we've got trouble.

Test Plan:
- I tested that this got rid of the test flakiness locally. I'm not sure
  how to easily write a test for this, because I don't actually know
  what the refcycle in the above is.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

zou3519 committed:
[Dynamo] Install module globals per output_graph

In tests, we ran into an issue where:
- In frame A, Dynamo would install a global
- We call reset()
- reset() did not delete the installed global due to a refcycle
- In frame B, Dynamo would re-use the same global
- Python gc ran, deleting the installed global, leading to the compiled
  version of frame B raising NameNotFound

This PR changes the following:
- module globals are now installed at a per-frame basis.
- renames install_global to install_global_unsafe: if the names are not
  unique and end up being re-used across frames, then we've got trouble.

Test Plan:
- I tested that this got rid of the test flakiness locally. I'm not sure
  how to easily write a test for this, because I don't actually know
  what the refcycle in the above is.

[ghstack-poisoned]

zou3519 committed:
Update on "[Dynamo] Install module globals per output_graph"


Fixes https://github.com/pytorch/pytorch/issues/117851

In tests, we ran into an issue where:
- In frame A, Dynamo would install a global
- We call reset()
- reset() did not delete the installed global due to a refcycle
- In frame B, Dynamo would re-use the same global
- Python gc ran, deleting the installed global, leading to the compiled
  version of frame B raising NameNotFound

This PR changes the following:
- module globals are now installed at a per-frame basis.
- renames install_global to install_global_unsafe: if the names are not
  unique and end up being re-used across frames, then we've got trouble.

Test Plan:
- I tested that this got rid of the test flakiness locally. I'm not sure
  how to easily write a test for this, because I don't actually know
  what the refcycle in the above is.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 committed:
Update on "[Dynamo] Install module globals per output_graph"


Fixes https://github.com/pytorch/pytorch/issues/117851

In tests, we ran into an issue where:
- In frame A, Dynamo would install a global
- We call reset()
- reset() did not delete the installed global due to a refcycle
- In frame B, Dynamo would re-use the same global
- Python gc ran, deleting the installed global, leading to the compiled
  version of frame B raising NameNotFound

This PR changes the following:
- module globals are now installed at a per-frame basis.
- renames install_global to install_global_unsafe: if the names are not
  unique and end up being re-used across frames, then we've got trouble.

Test Plan:
- I tested that this got rid of the test flakiness locally. I'm not sure
  how to easily write a test for this, because I don't actually know
  what the refcycle in the above is.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/output_graph.py:
```diff
@@ -1366,9 +1370,30 @@ def add_output_instructions(self, prefix: List[Instruction]) -> None:
 
+    def install_global_once(self, name, value) -> str:
+        """
+        Install a global once per frame being compiled. A frame graph breaking
+        into e.g. 2 graphs counts as two for the purposes of this function.
```
"install a global once per output_graph" makes this clearer, imo

voznesenskym commented on torch/_dynamo/output_graph.py:
```diff
@@ -1366,9 +1370,30 @@ def add_output_instructions(self, prefix: List[Instruction]) -> None:
+
+        Returns the name of the newly installed global.
+        """
+        name = f"{name}_c{self.compile_id}"
```
why do you need compile_id here?

zou3519 replied:
We are trying to prevent global sharing between different instances of torch.compile. I'm using "different output graphs" as the proxy for "different instances of torch.compile"

zou3519 replied:
The problem we ran into with test flakiness is that one output_graph would install `___module_torch_12345678`, another output_graph would rely on it, and then the first output_graph would die, taking `___module_torch_12345678` with it. Instead, each output_graph should have its own `___module_torch_12345678`, which is what compile_id is for.

ezyang commented on torch/_dynamo/output_graph.py:
```diff
@@ -1366,9 +1370,30 @@ def add_output_instructions(self, prefix: List[Instruction]) -> None:
+        a global installed by another instance. This can happen if we mangle
+        a global the same way across both instances.
+        """
         self.cleanups.append(CleanupHook.create(self.global_scope, name, value))
```
So is the idea we're going to eventually switch the rest of the sites over to this too?

zou3519 replied:
If we can, but they're difficult to refactor :/

ezyang reviewed (approved):
seems legit

zou3519 committed:
Update on "[Dynamo] Install module globals per output_graph"


Fixes https://github.com/pytorch/pytorch/issues/117851

In tests, we ran into an issue where:
- In frame A, Dynamo would install a global
- We call reset()
- reset() did not delete the installed global due to a refcycle
- In frame B, Dynamo would re-use the same global
- Python gc ran, deleting the installed global, leading to the compiled
  version of frame B raising NameNotFound

This PR changes the following:
- module globals are now installed at a per-frame basis.
- renames install_global to install_global_unsafe: if the names are not
  unique and end up being re-used across frames, then we've got trouble.

Test Plan:
- I tested that this got rid of the test flakiness locally. I'm not sure
  how to easily write a test for this, because I don't actually know
  what the refcycle in the above is.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 commented:
Just to understand: this PR does not fix the refcycle. It ensures that Dynamo generates different mangled names for different frames using compile_id counter. Is that correct?

zou3519 commented:
> Just to understand: this PR does not fix the refcycle. It ensures that Dynamo generates different mangled names for installed globals for different frames using compile_id counter. Is that correct?

Yes

zou3519 commented:
@pytorchbot merge -f "passed dynamo CI"
